File: pkg/dev_compiler/test/codegen_expected/language/dangling_else_test.js
dart_library.library('language/dangling_else_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__dangling_else_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const dangling_else_test = Object.create(null);
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  dangling_else_test.nestedIf1 = function(notTrue) {
    if (dart.test(notTrue)) return 'bad input';
    if (dart.test(notTrue)) {
      if (dart.test(notTrue)) {
        return 'bad';
      }
    } else {
      return 'good';
    }
    return 'bug';
  };
  dart.fn(dangling_else_test.nestedIf1, dynamicTodynamic());
  dangling_else_test.nestedIf2 = function(notTrue) {
    if (dart.test(notTrue)) return 'bad input';
    if (dart.test(notTrue)) {
      if (dart.test(notTrue)) {
        return 'bad';
      } else {
        if (dart.test(notTrue)) {
          return 'bad';
        }
      }
    } else {
      return 'good';
    }
    return 'bug';
  };
  dart.fn(dangling_else_test.nestedIf2, dynamicTodynamic());
  dangling_else_test.nestedWhile = function(notTrue) {
    if (dart.test(notTrue)) return 'bad input';
    if (dart.test(notTrue)) {
      while (dart.test(notTrue)) {
        if (dart.test(notTrue)) {
          return 'bad';
        }
      }
    } else {
      return 'good';
    }
    return 'bug';
  };
  dart.fn(dangling_else_test.nestedWhile, dynamicTodynamic());
  dangling_else_test.nestedFor = function(notTrue) {
    if (dart.test(notTrue)) return 'bad input';
    if (dart.test(notTrue)) {
      for (let i = 0; i < 3; i++) {
        if (i == 0) {
          return 'bad';
        }
      }
    } else {
      return 'good';
    }
    return 'bug';
  };
  dart.fn(dangling_else_test.nestedFor, dynamicTodynamic());
  dangling_else_test.nestedLabeledStatement = function(notTrue) {
    if (dart.test(notTrue)) return 'bad input';
    if (dart.test(notTrue)) {
      label:
        if (dart.test(notTrue)) {
          break label;
        }
    } else {
      return 'good';
    }
    return 'bug';
  };
  dart.fn(dangling_else_test.nestedLabeledStatement, dynamicTodynamic());
  dangling_else_test.main = function() {
    expect$.Expect.equals('good', dangling_else_test.nestedIf1(false));
    expect$.Expect.equals('good', dangling_else_test.nestedIf2(false));
    expect$.Expect.equals('good', dangling_else_test.nestedWhile(false));
    expect$.Expect.equals('good', dangling_else_test.nestedFor(false));
    expect$.Expect.equals('good', dangling_else_test.nestedLabeledStatement(false));
  };
  dart.fn(dangling_else_test.main, VoidTodynamic());
  // Exports:
  exports.dangling_else_test = dangling_else_test;
});

File: tests/language_2/if/dangling_else_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Tests dangling else. The VM should not have any problems, but dart2js or
// dart2dart could get this wrong.

import "package:expect/expect.dart";

nestedIf1(notTrue) {
  if (notTrue) return 'bad input';
  if (notTrue) {
    if (notTrue) {
      return 'bad';
    }
  } else {
    return 'good';
  }
  return 'bug';
}

nestedIf2(notTrue) {
  if (notTrue) return 'bad input';
  if (notTrue) {
    if (notTrue) {
      return 'bad';
    } else {
      if (notTrue) {
        return 'bad';
      }
    }
  } else {
    return 'good';
  }
  return 'bug';
}

nestedWhile(notTrue) {
  if (notTrue) return 'bad input';
  if (notTrue) {
    while (notTrue) {
      if (notTrue) {
        return 'bad';
      }
    }
  } else {
    return 'good';
  }
  return 'bug';
}

nestedFor(notTrue) {
  if (notTrue) return 'bad input';
  if (notTrue) {
    for (int i = 0; i < 3; i++) {
      if (i == 0) {
        return 'bad';
      }
    }
  } else {
    return 'good';
  }
  return 'bug';
}

nestedLabeledStatement(notTrue) {
  if (notTrue) return 'bad input';
  if (notTrue) {
    label:
    if (notTrue) {
      break label;
    }
  } else {
    return 'good';
  }
  return 'bug';
}

main() {
  Expect.equals('good', nestedIf1(false));
  Expect.equals('good', nestedIf2(false));
  Expect.equals('good', nestedWhile(false));
  Expect.equals('good', nestedFor(false));
  Expect.equals('good', nestedLabeledStatement(false));
}
